### ISSUE TITLE
side by side compare  tables when doing parameterized benchmarks

### ISSUE THREAD

**Issue by RonnyPfannschmidt** [ionelmc/pytest-benchmark#192]
in zokusei i compare different implementations of features using pytest-benchmark

with the standard output the long lists don't provide much of what iwant,
and id much rather see compact tables that pick specific values and line them up between test name and parameter value

a initial example is posted on https://gist.github.com/RonnyPfannschmidt/b0cd7a497bdfc69c1c72d8610acc4230

after some coordination with @ionelmc i'm happy to ustream this at a later point and will create a poc implementation in a conftest for zokusei as  a starter

**Comment by ionelmc** [comment]
So the problem is that the plugin has many options that control output and that won't work well with a compact table. For example:
* the compact table would have two grouping options (an additional one that controls the columns)
* the colums option doesn't have any sense anymore.

Just throwing some ideas here.

What if I would add a hook that allows users to completely override the output parts? Would it be possible  for that hook to exist in a 3rd party package that you can pipinstall? Eg: `pip install pytest-benchmark-compact-table` and then do some hook/plugin magic.

Or maybe pytest-benchmark could enumerate some setuptools entrypoint and then you use `--pytest-benchmark-output=compact-table` (default would be `--pytest-benchmark-output=builtin`)?

@RonnyPfannschmidt I'm open to suggestions but I believe the correct way is to enable an external implementation.

**Comment by RonnyPfannschmidt** [comment]
Yes, having customizable output for certain groups seems to be a good way to do it

I believe it may also open the door to ship some of them with the plugin after initially experimenting outside 

I believe it should be possible to have multiple outputs /groups at once 

But the ux/api on that is unclear

**Comment by ionelmc** [comment]
> Yes, having customizable output for *certain groups* seems to be a good way to do it

Ok, I haven't considered that. Well I guess I need to have a way to attach some metadata to a group. But a group I created based on what group option and/or what group name you have set for each test.

For that I could imagine a couple ways:
* Set group metadata on a test, and do some conflict checks for situations where multiple tests in the same group set same option to different values. Not really ideal, and confusing to use.
* Set group metadata via a hook. Tho this allows users to set metadata for groups that do not exist, or aren't used.
* Some other global configuration thingy?

Not sure what's best here, tho I think the first one is the worst. Any other ideas?

A hook could look like:

```python

def report_progress(iterable, terminal_reporter, format_string, **kwargs):
    ...

class AbstractBenchmarkGroup(ABC):
    def display(self, tr: TerminalReporter, columns: List[str], progress_reporter=report_progress):
        # columns would be list of property names like avg, min, max etc
        ...

def pytest_benchmark_group_factory(name: str, benchmarks: List):
    return BuiltinTabularBenchmarkGroup(name, benchmarks)
```

Still doesn't feel right - if you'd implement a CompactTabularBenchmarkGroup the `columns` argument would go unused and you'd need a different way to tell where the colums come from since the columns is a list of those stat names.

Maybe if we'd allow something like `--benchmark-columns=param:backend` and auto switch to a compact table would be better from an user's perspective?

**Comment by RonnyPfannschmidt** [comment]
I think this need brainstorming and yagni considerations a like, if you like we can have a look at doing a little voice/video call to explore the problem domain

Alternatively we ought to spec this in more detail

**Comment by ionelmc** [comment]
Sure, preferred time? Anyone else if you want to jump in we gonna be in pytest-dev's discord: https://discord.gg/qqPeCEDvBt

**Comment by RonnyPfannschmidt** [comment]
Today is bad for me, let's figure something until the weekend

**Comment by ionelmc** [comment]
Alright, so I'm thinking now to also make storage pluggable in the same way - https://github.com/ionelmc/pytest-benchmark/issues/200 - so for reporting it could be like this:

```python
    entry_points={
        'pytest_benchmark': [
            'cli-add-options = pytest_benchmark_fancypants:add_options',
            'result-class = pytest_benchmark_fancypants:FancyResults',
        ],
    }
```

FancyResults would have the same interface as TableResults mostly - I'd need to do some refactoring to allow it to take the custom options (that one could add via cli-add-options).

Another thing to deal with I guess would be what would happen with multiple plugins implementing result-class entrypoint. I'd probably need to add a `--benchmark-reporter` option (that would work similar to `--benchmark-storage`) to make this nice. So I guess we'd end up with this?


```python
    entry_points={
        'pytest_benchmark': [
            'result-uri-example = pytest_benchmark_fancypants:uri_example',
            'result-uri-prefix = pytest_benchmark_fancypants:uri_prefix',
            'result-uri-parse = pytest_benchmark_fancypants:uri_parse',
            'cli-add-options = pytest_benchmark_fancypants:add_options',
        ],
    }
```

Tho this will allow passing in options both via custom cli options and uri value - I don't really like this but at least it's consistent with the proposed storage entrypoint api I guess?

@RonnyPfannschmidt what do you think?

@RonnyPfannschmidt any other ideas

**Comment by RonnyPfannschmidt** [comment]
Added this to the thinking backlog, currently I'm a bit stretched, it may take a few days to get back to this in a sensible way

**Comment by RonnyPfannschmidt** [comment]
hmm, currently im stuck wth something else again, i hope to get to the gist of it next week
